Data/ByteString/Lazy/Internal.hs

```diff
@@ -230,27 +231,29 @@ eq :: ByteString -> ByteString -> Bool
 eq Empty Empty = True
 eq Empty _     = False
 eq _     Empty = False
-eq (Chunk a@(S.BS ap al) as) (Chunk b@(S.BS bp bl) bs) =
-  case compare al bl of
-    LT -> a == S.BS bp al && eq as (Chunk (S.BS (S.plusForeignPtr bp al) (bl - al)) bs)
-    EQ -> a == b && eq as bs
-    GT -> S.BS ap bl == b && eq (Chunk (S.BS (S.plusForeignPtr ap bl) (al - bl)) as) bs
+eq ac@(Chunk a@(S.BS ap al) as) bc@(Chunk b@(S.BS bp bl) bs)
+  | isTrue# (reallyUnsafePtrEquality# ac bc) = True
+  | otherwise = case compare al bl of
+      LT -> a == S.BS bp al && eq as (Chunk (S.BS (S.plusForeignPtr bp al) (bl - al)) bs)
+      EQ -> a == b && eq as bs
+      GT -> S.BS ap bl == b && eq (Chunk (S.BS (S.plusForeignPtr ap bl) (al - bl)) as) bs
```

Comment:
In the `LT` and `GT` cases, we'd recursively run the pointer equality checks on freshly allocated `Chunk`s – which is totally wasteful

It _might_ be better to call an "inner" function here which doesn't perform the pointer equality check.

`cmp` has the same issue.